clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -1223,6 +1223,11 @@ private void ensureValidRecordSize(int size) {
      */
     @Override
     public void flush() {
+        if (Thread.currentThread() == this.ioThread) {
+            log.error("flush invocation detected in the callback. This can cause a deadlock due to thread blocking.");
```

Comment:
I would change the wording here because actually the code is preventing the call and that means there is no risk of a deadlock. Something like "KafkaProducer.flush() invocation detected inside callback. This is not permitted because of the risk of deadlock."

The message in the exception could be the same. The choice of KafkaException seems most appropriate to me.

Reply:
I would like to know if this log is necessary? In my opinion, this information is already recorded bt `KafkaException`.

Reply:
I think it's ok to use both logging and exceptions simultaneously, as the user might want immediate hints displayed on the console.